tests/integration/resource_aware_pool_tests.cpp

```diff
@@ -0,0 +1,156 @@
+  int* compare_ptr1 = ptr;
+
+  EXPECT_EQ(getResource(pool, ptr), r1);
+  EXPECT_EQ(getPendingSize(pool), 0);
+
+  host_sleep(ptr);
+
+  pool.deallocate(r1, ptr);
+  // EXPECT_EQ(getPendingSize(pool), 0); // When only using host, there will be no pending chunks
+
+  ptr = static_cast<int*>(pool.allocate(r2, 2048));
+  int* compare_ptr2 = ptr;
+
+  EXPECT_TRUE(r1 == r2);
+  EXPECT_EQ(compare_ptr1, compare_ptr2); // only 1 host resource available, no possible data race
+
+  pool.deallocate(r2, ptr);
+}
+
+#if defined(UMPIRE_ENABLE_CUDA) || defined(UMPIRE_ENABLE_HIP)
+
+using clock_value_t = long long;
+
+__device__ clock_value_t my_clock()
+{
+  return clock64();
+}
+
+__device__ void my_sleep(clock_value_t sleep_cycles)
+{
+  clock_value_t start = my_clock();
+  clock_value_t cycles_elapsed;
+  do {
+    cycles_elapsed = my_clock() - start;
+  } while (cycles_elapsed < sleep_cycles);
+}
+
+__global__ void do_sleep(double* ptr)
+{
+  int id = blockIdx.x * blockDim.x + threadIdx.x;
+  ptr[id] = id * 1.0;
+  my_sleep(1000000);
+  ptr[id] = id * 2.0;
+}
+
+std::vector<std::string> get_allocator_strings()
+{
+  std::vector<std::string> allocators;
+
+  allocators.push_back("DEVICE");
+  // auto& rm = umpire::ResourceManager::getInstance();
+  // for (int id = 0; id < rm.getNumDevices(); id++) {
+  // allocators.push_back(std::string{"DEVICE::" + std::to_string(id)});
+  //}
+#if defined(UMPIRE_ENABLE_UM)
+  allocators.push_back("UM");
+#endif
+#if defined(UMPIRE_ENABLE_CONST)
+  allocators.push_back("DEVICE_CONST");
+#endif
+#if defined(UMPIRE_ENABLE_PINNED)
+  allocators.push_back("PINNED");
+#endif
+
+  return allocators;
+}
+
+class ResourceAwarePoolTest : public ::testing::TestWithParam<std::string> {
+ public:
+  virtual void SetUp()
+  {
+    auto& rm = umpire::ResourceManager::getInstance();
+    m_pool = rm.makeAllocator<umpire::strategy::ResourceAwarePool>(std::string{"rap-pool-" + GetParam()},
+                                                                   rm.getAllocator(GetParam()));
+  }
+
+  virtual void TearDown()
+  {
+    m_pool.release();
+  }
+
+  umpire::Allocator m_pool;
+};
+
+TEST_P(ResourceAwarePoolTest, Check_States)
+{
+#if defined(UMPIRE_ENABLE_CUDA)
+  Cuda d1, d2;
+#elif defined(UMPIRE_ENABLE_HIP)
+  Hip d1, d2;
+#endif
+
+  Resource r1{d1}, r2{d2};
+
+  double* ptr = static_cast<double*>(m_pool.allocate(r1, 1024));
+
+  EXPECT_EQ(getResource(m_pool, ptr), r1);
+  EXPECT_EQ(getPendingSize(m_pool), 0);
+
+#if defined(UMPIRE_ENABLE_CUDA)
+  do_sleep<<<1, 32, 0, d1.get_stream()>>>(ptr);
+#elif defined(UMPIRE_ENABLE_HIP)
+  hipLaunchKernelGGL(do_sleep, 1, 32, 0, d1.get_stream(), ptr);
+#endif
+
+  m_pool.deallocate(r1, ptr);
+  EXPECT_EQ(getPendingSize(m_pool), 1);
+
+  double* ptr2 = static_cast<double*>(m_pool.allocate(r2, 2048));
+
+  EXPECT_FALSE(r1 == r2);
+  EXPECT_EQ(getResource(m_pool, ptr2), r2);
+  EXPECT_NE(ptr, ptr2); // multiple device resources, possible data race, needs different addr
```

Comment:
Can you add a test with multiple device resources where it can reuse the same pointer?

Reply:
When I reallocate a ptr with the same resource, it will reuse the same pointer. I added a case where there is d1 and d2, but d1 is used to create ptr and ptr2 (they are now the same) and then d2 is used to create ptr3 (this one will be different)

Reply:
I meant a case where you allocate on one resource, run a kernel on that resource, deallocate on that resource, then explicitly synchronize that resource. At that point, the freed block should no longer be pending and if you allocate on a second resource, you should be able to reuse it.

Reply:
I see. I can test that...

Reply:
But do you want to check the state of all the pending events if there is already free memory that could serve the request?

Reply:
Added to test, this currently works - it reuses the pointer

Reply:
Jason, if I understand what you're asking, we currently prioritize reusing memory that is pending rather than using a new chunk.. the first pending check we find that will work, we reuse.